include/fiction/algorithms/simulation/sidb/displacement_robustness_domain.hpp

```diff
@@ -0,0 +1,556 @@
+                                   stats.num_operational_sidb_displacements);
+    }
+
+  private:
+    /**
+     * This function calculates all permitted displacements for each SiDB based on the specified allowed displacements.
+     */
+    void calculate_permitted_displacements_for_each_sidb() noexcept
+    {
+        std::vector<std::vector<cell<Lyt>>> all_possible_sidb_misplacements = {};
+        all_possible_sidb_misplacements.reserve(layout.num_cells());
+        layout.foreach_cell(
+            [&](const auto& c)
+            {
+                if constexpr (has_siqad_coord_v<Lyt>)
+                {
+                    auto new_pos_se = siqad::to_fiction_coord<cube::coord_t>(c);
+                    auto new_pos_nw = siqad::to_fiction_coord<cube::coord_t>(c);
+                    // the cell c is not a fixed cell, i.e., displacement is considered.
+                    if (params.fixed_sidbs.find(c) == params.fixed_sidbs.cend())
+                    {
+                        new_pos_se.x -= static_cast<decltype(new_pos_se.x)>(params.displacement_variations.first);
+                        new_pos_nw.x += static_cast<decltype(new_pos_nw.x)>(params.displacement_variations.first);
+                        if (params.allow_dimer_change_in_y_direction)
+                        {
+                            new_pos_nw.y = siqad::to_fiction_coord<cube::coord_t>(siqad::coord_t{c.x, c.y, 0}).y;
+                            new_pos_se.y = siqad::to_fiction_coord<cube::coord_t>(siqad::coord_t{c.x, c.y, 1}).y;
+                        }
+                        new_pos_se.y -= static_cast<decltype(new_pos_se.y)>(params.displacement_variations.second);
+                        new_pos_nw.y += static_cast<decltype(new_pos_nw.y)>(params.displacement_variations.second);
+                    }
+                    const auto all_coord = all_coordinates_in_spanned_area<cell<Lyt>>(
+                        siqad::to_siqad_coord(new_pos_se), siqad::to_siqad_coord(new_pos_nw));
+                    all_possible_sidb_misplacements.push_back(all_coord);
+                }
+                else
+                {
+                    auto new_pos_se = c;
+                    auto new_pos_nw = c;
+                    if (params.fixed_sidbs.find(c) == params.fixed_sidbs.cend())
+                    {
+                        new_pos_se.x -= params.displacement_variations.first;
+                        new_pos_se.y -= params.displacement_variations.second;
+                        new_pos_nw.x += params.displacement_variations.first;
+                        new_pos_nw.y += params.displacement_variations.second;
+                    }
+
+                    const auto all_coord = all_coordinates_in_spanned_area<cell<Lyt>>(new_pos_se, new_pos_nw);
+                    all_possible_sidb_misplacements.push_back(all_coord);
+                }
+            });
+        all_possible_sidb_displacements = all_possible_sidb_misplacements;
+    }
+    /**
```

Comment:
The code is fine it just looks a bit cramped. Inserting some blank lines gives it room to breathe ;)